### DISCUSSION TITLE
Proposal: Readonly tuples and covariance

### DISCUSSION THREAD

**Discussion by MkazemAkhgary** [dotnet/csharplang#1298] [Language Ideas]
using `in` keyword we could define read-only members for tuples.

     (in Foo foo, Bar bar) t = (x, y);
     t.foo = a; // error
     t.bar = b; // ok

only declared tuples can have read-only members. multiple assignments or deconstructions cant have `in` modifier.

     (in Foo foo, Bar bar) = (x, y); // invalid because tuple is not declared.

read-write tuples can be assigned to tuple with read-only members and vice versa.

read-only tuples could also form covariance structures.

```
public interface IFoo<TKey, out TValue> : IEnumerable<(TKey, in TValue)>
{
    // TValue is convariance because its read-only.
}
```

currently tuples does not support covariance. it would be nice to have them support this feature. alternatives like using struct and interface causes boxing. using class causes heap allocation.

Whether this requires CLR change or not is not relevant to me. we had this discussion before. I'm tired of this.

I'm just proposing a syntax, what it should look like to have read only tuples. (with covariance behind)

**Comment by gafter** [comment]
Syntax looks OK. We like to have semantics too. Currently tuples are mapped to an instance is `System.ValueTuple`. Do you have a mapping in mind for these types? Or are you satisfied to have this proposal remain at this stage?

**Comment by Joe4evr** [comment]
Couple of observations:
1. There's no way this can be applied at runtime with the current tuple types
2. Using tuples as part of your public API surface is highly discouraged (you should have a dedicated type instead because there's no telling if you are going to add/remove members in the future)

**Comment by MkazemAkhgary** [comment]
>Do you have a mapping in mind for these types? 

@gafter it seems like either I don't have knowledge for that or I don't understand what is meant by term "mapping" (which again brings lack of my knowledge). but I would appreciate to see active discussion or some work on this.

@Joe4evr 

1- I see. doesn't sound good.
2- yes. but for this specific case though there is telling that members never change. because all that is, is key and value. I could use `KeyValuePair`.

**Comment by ufcpp** [comment]
looks the same as your previous proposals
https://github.com/dotnet/csharplang/issues/1152
https://github.com/dotnet/csharplang/issues/1156

**Comment by MkazemAkhgary** [comment]
@ufcpp Thanks for notice. ill go ahead and close previous ones.

**Comment by Mafii** [comment]
The Mapping goes hand in hand with what @Joe4evr mentioned. It will need a type/mapping that does for readonly tuples (and mixed readonly/non-readonly tuples), what `System.ValueTuple` does for Tuples currently.

This type will have to support readonly types. This is where this gets confusing/hard.

There has to be a exact type for each readonly/non-readonly tuple combination. This is 2^amount of properties. Currently we only need the amount of properties. This is, if done like today, 2^8=256 types (system.valuetuple contains up to 8 fields via Create syntax (https://msdn.microsoft.com/en-us/library/system.valuetuple(v=vs.110).aspx) and supports even more IIRC.

Are my assumptions correct @gafter ?

**Comment by HaloFour** [comment]
@MkazemAkhgary 

I don't see there being much point in this.  Structs being value types are copied by value so the mutability of the fields are effectively irrelevant.  There were long conversations around this when tuples were being designed that the fact that tuples are structs making their mutability not a problem.

[Proposal: Language support for Tuples](https://github.com/dotnet/roslyn/issues/347)

> ### Mutability
> Should tuples be mutable or immutable? The nice thing about them being structs is that the user can choose. If a reference to the tuple is readonly then the tuple is readonly.
> 
> Now a local variable cannot be readonly, unless we adopt [#115](https://github.com/dotnet/roslyn/issues/115) (which is likely), but that isn't too big of a deal, because locals are only used locally, and so it is easier to stick to an immutable discipline if you so choose.
>
> If tuples are used as fields, then those fields can be readonly if desired.

Also, nothing about this proposal would help with variance.  Structs can't be variant, whether or not their fields are readonly.

**Comment by quinmars** [comment]
I think #188 should be enough:

```csharp
 readonly (Foo foo, Bar bar) t = (x, y);
 t.foo = a; // error
 t.bar = b; // error
```
BTW: `in` now a days stands for `readonly ref`. It'd be confusing if it would be used here for a non `ref` type.
